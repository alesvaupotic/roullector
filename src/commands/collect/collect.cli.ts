import {
  Command,
  Option,
} from 'commander';

import { collect } from '$commands/collect/collect';
import {
  defaultCliCollectOptions,
  keyTransformCliToFunc,
} from '$commands/collect/collect.constants';
import type {
  CollectOptions,
  CliCollectOptions,
} from '$commands/collect/collect.types';

export function collectCli() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-i, --inDir <path>',
      'directory path to collect',
      defaultCliCollectOptions.inDir
    )
    .option(
      '-e, --extensions <list>',
      'comma-separated list of file extensions to collect',
      defaultCliCollectOptions.extensions,
    )
    .option(
      '-x, --ignorePatterns <list>',
      'comma-separated list of filename patterns (string literal or regex) to ignore',
      defaultCliCollectOptions.ignorePatterns,
    )
    .option(
      '--no-output',
      'don\'t write to disk but print to console instead',
      defaultCliCollectOptions.output,
    )
    .option(
      '-o, --outDir <path>',
      'outputs generated files to this folder',
      defaultCliCollectOptions.outDir,
    )
    .option(
      '-d, --depth <number>',
      'depth of inDir to collect',
      defaultCliCollectOptions.depth,
    )
    .addOption(
      new Option( '-t, --keyTransform <variant>', 'how to transform route key')
        .choices(['none', 'camelCase'])
        .default(defaultCliCollectOptions.keyTransform)
    )
    .option(
      '-k, --dirkey <name>',
      'how to transform route key',
      defaultCliCollectOptions.dirkey,
    )
    .option(
      '--no-utils',
      'don\'t generate utils for building path with arguments',
      defaultCliCollectOptions.utils,
    )
    .option(
      '--no-typescript',
      'outputs files in javascript instead of typescript',
      defaultCliCollectOptions.typescript,
    )
    .option(
      '--verbose',
      'prints more info during operation',
      defaultCliCollectOptions.verbose,
    )
    .action((rawOptions: CliCollectOptions) => {
      const options: CollectOptions = {
        ...rawOptions,
        extensions: rawOptions.extensions?.split(',') ?? [],
        ignorePatterns: rawOptions.ignorePatterns?.split(',') ?? [],
        depth: Number(rawOptions.depth) || Infinity,
        keyTransform: keyTransformCliToFunc[rawOptions.keyTransform],
      };
      if (options.verbose) {
        console.info('Running "collect" with options:');
        console.table(Object.fromEntries(
          Object.entries(rawOptions).map(
            ([key, value]) => ([
              key,
              { value },
            ])
          )),
        );
      }

      const output = collect(options);
      if (!options.output) {
        console.log(output);
      }
    });
}
