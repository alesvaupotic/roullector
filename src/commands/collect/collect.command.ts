import {
  statSync,
  readdirSync,
  writeFileSync,
  mkdirSync,
  existsSync,
  rmSync,
} from 'fs';
import {
  basename,
  join,
  resolve,
} from 'path';

import type { CollectOptions } from './collect.types';

export type DeepObject<T> = {
  [key: string]: T | DeepObject<T>;
};

function extractRouteMapping(options: CollectOptions, inputDir: string = options.inDir, currentPath = '', level = 0): DeepObject<string> {
  const children = readdirSync(inputDir);

  const map: DeepObject<string> = {};

  if (level <= options.depth) {
    for (const child of children) {
      const absPath = join(inputDir, child);
      const stats = statSync(absPath);

      // skip if any ignore pattern matches
      const filename = basename(child);
      if (options.ignorePatterns.some(pattern => new RegExp(pattern).test(filename))) continue;

      if (stats.isDirectory()) {
        const nextPath = `${currentPath}/${child}`;
        const childMap = extractRouteMapping(options, absPath, nextPath, level + 1);
        if (Object.keys(childMap).length) {
          if (!childMap['index']) {
            childMap[options.dirkey] = nextPath;
          }
          map[options.keyTransform(child)] = childMap;
        }
      } else {
        // skip if extension doesn't match
        if (options.extensions.every((ext) => !filename.endsWith(ext))) continue;

        // transform filename to route key
        let routeKey = filename.replace(/\.[^/.]+$/, ''); // trim extension
        routeKey = options.keyTransform(routeKey); // by default camelCase-ify key

        let routeName = routeKey;
        if (routeKey === 'index') {
          // no explicit `index`
          routeName = '';
        }
        let nextPath = currentPath;
        if (routeName) {
          nextPath += `/${routeName}`;
        }

        if (!nextPath && level === 0) {
          // root level index route
          nextPath = '/';
        }

        map[routeKey] = nextPath;
      }
    }
  }

  return map;
}

export function generateJSON(options: CollectOptions) {
  const map = extractRouteMapping(options);
  const outputPath = resolve(options.outDir, 'routes.json');
  writeFileSync(
    outputPath,
    JSON.stringify(map, null, 2), // spacing level = 2
    { encoding: 'utf-8' },
  );
}

// export function generateUtils() {

// }

export function collect(options: CollectOptions) {
  const { outDir } = options;
  try {
    if (!existsSync(outDir)) {
      mkdirSync(outDir, { recursive: true });
    }
    generateJSON(options);
  } catch (error) {
    console.error(error);
    const data = readdirSync(outDir);
    if (data.length === 0) {
      rmSync(outDir, { recursive: true });
    }
  }
}
