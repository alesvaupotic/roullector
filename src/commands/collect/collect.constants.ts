import type {
  CliCollectOptions,
  CollectOptions,
} from './collect.types';

export function CamelCasify(str: string) {
  return str
    .trim()
    .replace(/([-_\s]+.)/g, function($1) {
      return $1.replace(/[-_\s]/g, '').toUpperCase();
    })
    .replace(/^(.)/, function($1) {
      return $1.toLowerCase();
    });
}

export const defaultCollectOptions: CollectOptions = {
  inDir: 'src/routes',
  extensions: ['.svelte'],
  ignorePatterns: ['^_'],
  outDir: 'src/generated/routing',
  typescript: true,
  keyTransform: CamelCasify,
  verbose: false,
  depth: Infinity,
  dirkey: '__dir',
  utils: true,
};

export const defaultCliCollectOptions: CliCollectOptions = {
  ...defaultCollectOptions,
  extensions: defaultCollectOptions.extensions.join(','),
  ignorePatterns: defaultCollectOptions.ignorePatterns.map((pattern) => (pattern instanceof RegExp) ? pattern.source : pattern).join(','),
  depth: 'Infinity',
  keyTransform: 'camelCase',
};

export const keyTransformCliToFunc: Record<CliCollectOptions['keyTransform'], (str) => string> = {
  camelCase: CamelCasify,
};
