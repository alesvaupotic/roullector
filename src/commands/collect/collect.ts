import {
  readdirSync,
  mkdirSync,
  existsSync,
  rmdirSync,
} from 'fs';

import type {
  CollectOptions,
  CollectOutput,
} from '$commands/collect/collect.types';
import { generateJSON } from '$commands/collect/generators/json';
import { generateRouteUtil } from '$commands/collect/generators/route';

export function collect(options: CollectOptions): CollectOutput {
  const { outDir } = options;
  const output = {
    json: null,
    route: null,
  };
  try {
    if (!existsSync(outDir)) {
      mkdirSync(outDir, { recursive: true });
    }
    output.json = generateJSON(options);
    if (options.utils) {
      output.route = generateRouteUtil(options);
    }
  } catch (error) {
    console.error(error);
    const data = readdirSync(outDir);
    if (data.length === 0) {
      rmdirSync(outDir, { recursive: true });
    }
    throw error;
  }
  return output;
}
