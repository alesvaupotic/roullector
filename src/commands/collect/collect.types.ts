type Modify<T, R> = Omit<T, keyof R> & R;

export type CollectOptions = {
  /** input directory path to collect route data from */
  inDir: string;
  /** file extensions */
  extensions: string[];
  /** patterns to ignore files */
  ignorePatterns: (string|RegExp)[];
  /** route data output directory path */
  outDir: string;
  /** prints more info during operation */
  verbose: boolean;
  /** how to transform route key */
  keyTransform: (key: string) => string;
  /** depth of inDir to collect */
  depth: number;
  /** key to save path for directories with no index file */
  dirkey: string;
  /** generate utils for building path with arguments */
  utils: boolean;
  /** whether to output files in typescript */
  typescript: boolean;
};

export type CliCollectOptions = Modify<CollectOptions, {
  depth: string;
  extensions?: string;
  ignorePatterns?: string;
  keyTransform: 'camelCase';
}>;
