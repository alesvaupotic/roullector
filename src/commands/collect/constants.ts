import type { CollectOptions } from './types';

export const defaultCollectOptions: CollectOptions = {
  pattern: 'src/routes',
  outDir: 'src/generated/routing',
  typescript: true,
};
