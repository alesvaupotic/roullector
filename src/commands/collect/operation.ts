// import { statSync } from 'fs';
// import { dirname } from 'path';

import { sync as globSync } from 'glob';

import type { CollectOptions } from './types';

export type DeepObject<T> = {
  [key: string]: T | DeepObject<T>;
};

export function generateJSON(pattern: string) {
  const files = globSync(pattern);
  for (const file of files) {
    console.log(file);
  }
}

// export function generateUtils() {

// }

export function operation(options: CollectOptions) {
  const { pattern } = options;
  generateJSON(pattern);
}
