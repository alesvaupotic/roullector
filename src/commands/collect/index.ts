import { Command } from 'commander';

import { defaultCollectOptions } from './constants';
import type { CollectOptions } from './types';

export function collect() {
  return new Command('collect')
    .description('collect route data')
    .option(
      '-p, --pattern <type>',
      'glob pattern to search for files',
      defaultCollectOptions.pattern
    )
    .option(
      '-o, --outDir <type>',
      'outputs generated files to this folder',
      defaultCollectOptions.outDir
    )
    .option(
      '-t, --typescript',
      'outputs generated files in typescript',
      defaultCollectOptions.typescript
    )
    .action((rawOptions: CollectOptions) => {
      const options = {
        ...defaultCollectOptions,
        ...rawOptions,
      };
    });
}
