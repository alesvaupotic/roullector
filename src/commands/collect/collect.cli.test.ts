import { resolve } from 'path';

import {
  vol,
  fs as memfs,
} from 'memfs';

import { collectCli } from '$commands/collect/collect.cli';
import { FILES } from '$tests/fixtures/memfs';

import { defaultCollectOptions } from './collect.constants';

jest.mock('fs', () => {
  const fs = jest.requireActual('fs');
  const sourcePath = './generators/route/source.ts';
  const sourceOfRouteUtil = fs.readFileSync(resolve(__dirname, sourcePath), 'utf-8');
  vol.fromJSON({
    [sourcePath]: sourceOfRouteUtil,
  }, __dirname);

  return memfs;
});

const savedJSON = vol.toJSON();
beforeEach(() => {
  vol.fromNestedJSON(FILES, defaultCollectOptions.inDir);
  vol.fromJSON(savedJSON);
});
afterEach(() => {
  vol.reset();
});


test('cli collect: default --no-output', () => {
  const writeSpy = jest.spyOn(process.stdout, 'write').mockImplementation();
  const program = collectCli().exitOverride();
  program.parse(['node', 'collect', '--no-output']);
  expect(writeSpy).toHaveBeenCalledTimes(1);
  expect(writeSpy.mock.calls[writeSpy.mock.calls.length - 1][0]).toMatchSnapshot('cli collect: default');
});

test('cli collect: verbose', () => {
  const logInfo = jest.spyOn(console, 'info').mockImplementation();
  const tableSpy = jest.spyOn(console, 'table').mockImplementation();
  const program = collectCli().exitOverride();
  program.parse(['node', 'collect', '--no-output', '--verbose']);
  expect(logInfo).toHaveBeenCalled();
  expect(tableSpy).toHaveBeenCalled();
});
