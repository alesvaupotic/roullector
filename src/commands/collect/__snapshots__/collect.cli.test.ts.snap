// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli collect: default --no-output: cli collect: default 1`] = `
"{
  \\"json\\": \\"{\\\\n  \\\\\\"admin\\\\\\": {\\\\n    \\\\\\"users\\\\\\": {\\\\n      \\\\\\"[id]\\\\\\": {\\\\n        \\\\\\"index\\\\\\": \\\\\\"/admin/users/[id]\\\\\\"\\\\n      },\\\\n      \\\\\\"assets\\\\\\": {\\\\n        \\\\\\"index\\\\\\": \\\\\\"/admin/users/assets\\\\\\",\\\\n        \\\\\\"photos\\\\\\": {\\\\n          \\\\\\"[id]\\\\\\": {\\\\n            \\\\\\"index\\\\\\": \\\\\\"/admin/users/assets/photos/[id]\\\\\\"\\\\n          },\\\\n          \\\\\\"__dir\\\\\\": \\\\\\"/admin/users/assets/photos\\\\\\"\\\\n        },\\\\n        \\\\\\"posts\\\\\\": {\\\\n          \\\\\\"youtube\\\\\\": {\\\\n            \\\\\\"[id]\\\\\\": \\\\\\"/admin/users/assets/posts/youtube/[id]\\\\\\",\\\\n            \\\\\\"[slug]\\\\\\": \\\\\\"/admin/users/assets/posts/youtube/[slug]\\\\\\",\\\\n            \\\\\\"__dir\\\\\\": \\\\\\"/admin/users/assets/posts/youtube\\\\\\"\\\\n          },\\\\n          \\\\\\"__dir\\\\\\": \\\\\\"/admin/users/assets/posts\\\\\\"\\\\n        }\\\\n      },\\\\n      \\\\\\"__dir\\\\\\": \\\\\\"/admin/users\\\\\\"\\\\n    },\\\\n    \\\\\\"__dir\\\\\\": \\\\\\"/admin\\\\\\"\\\\n  },\\\\n  \\\\\\"index\\\\\\": \\\\\\"/\\\\\\",\\\\n  \\\\\\"me\\\\\\": \\\\\\"/me\\\\\\",\\\\n  \\\\\\"signIn\\\\\\": \\\\\\"/signIn\\\\\\",\\\\n  \\\\\\"signUp\\\\\\": \\\\\\"/signUp\\\\\\"\\\\n}\\",
  \\"route\\": \\"export { default as AppRoutes } from './routes.json';\\\\n/**\\\\n * build a complete path with injected arguments\\\\n * @param path {string} based path\\\\n * @param args {string[]} arguments to inject\\\\n */\\\\nexport function route(path: string, ...args: string[]): string {\\\\n  const params = path.match(/\\\\\\\\[[a-zA-Z_-]+\\\\\\\\]/g) ?? [];\\\\n  for (const i in params) {\\\\n    path = path.replace(params[i], args[i]);\\\\n  }\\\\n  return path;\\\\n}\\\\n\\\\n\\"
}"
`;
